clients/venice-samza/src/main/java/com/linkedin/venice/samza/VeniceSystemProducer.java

```diff
@@ -366,45 +401,56 @@ public synchronized void start() {
-     * version in the code base is registered in Venice backend; if not, fail fast in start phase before start writing
-     * Kafka messages that Venice backend couldn't deserialize.
-     */
-    if (verifyLatestProtocolPresent) {
-      LOGGER.info("Start verifying the latest protocols at runtime are valid in Venice backend.");
-      // Discover the D2 service name for the system store
-      String kafkaMessageEnvelopSchemaSysStore = AvroProtocolDefinition.KAFKA_MESSAGE_ENVELOPE.getSystemStoreName();
-      D2ServiceDiscoveryResponse sysStoreDiscoveryResponse = (D2ServiceDiscoveryResponse) controllerRequestWithRetry(
-          () -> D2ControllerClient.discoverCluster(
-              primaryControllerColoD2Client,
-              primaryControllerD2ServiceName,
-              kafkaMessageEnvelopSchemaSysStore),
-          2);
-      ClientConfig clientConfigForKafkaMessageEnvelopeSchemaReader =
-          ClientConfig.defaultGenericClientConfig(kafkaMessageEnvelopSchemaSysStore);
-      clientConfigForKafkaMessageEnvelopeSchemaReader.setD2ServiceName(sysStoreDiscoveryResponse.getD2Service());
-      clientConfigForKafkaMessageEnvelopeSchemaReader.setD2Client(childColoD2Client);
-      SchemaReader kafkaMessageEnvelopeSchemaReader =
-          ClientFactory.getSchemaReader(clientConfigForKafkaMessageEnvelopeSchemaReader, null);
-      new SchemaPresenceChecker(kafkaMessageEnvelopeSchemaReader, AvroProtocolDefinition.KAFKA_MESSAGE_ENVELOPE)
-          .verifySchemaVersionPresentOrExit();
-      LOGGER.info("Successfully verified the latest protocols at runtime are valid in Venice backend.");
-    }
+    final TransportClient transportClient;
+    if (discoveryUrl.isPresent()) {
+      this.controllerClient =
+          ControllerClient.discoverAndConstructControllerClient(storeName, discoveryUrl.get(), sslFactory, 1);
+      transportClient = new HttpTransportClient(discoveryUrl.get());
+    } else {
+      this.primaryControllerColoD2Client = getStartedD2Client(primaryControllerColoD2ZKHost);
+      this.childColoD2Client = getStartedD2Client(veniceChildD2ZkHost);
+
+      // Discover cluster
+      D2ServiceDiscoveryResponse discoveryResponse = (D2ServiceDiscoveryResponse) controllerRequestWithRetry(
+          () -> D2ControllerClient
+              .discoverCluster(primaryControllerColoD2Client, primaryControllerD2ServiceName, this.storeName),
+          10);
+      String clusterName = discoveryResponse.getCluster();
+      LOGGER.info("Found cluster: {} for store: {}", clusterName, storeName);
+
+      /**
+       * Verify that the latest {@link com.linkedin.venice.serialization.avro.AvroProtocolDefinition#KAFKA_MESSAGE_ENVELOPE}
+       * version in the code base is registered in Venice backend; if not, fail fast in start phase before start writing
+       * Kafka messages that Venice backend couldn't deserialize.
+       */
+      if (verifyLatestProtocolPresent) {
```

Comment:
Hmm... making this check execute only when `discoveryUrl` is not present does not seem ideal.

@nisargthakkar, does this check (or cluster disco, for that matter) really need to depend on D2?

Reply:
@FelixGV @nisargthakkar 

It kind of works like
```java
    final TransportClient transportClient;
    if (discoveryUrl.isPresent()) {
      this.controllerClient =
          ControllerClient.discoverAndConstructControllerClient(storeName, discoveryUrl.get(), sslFactory, 1);

      /**
       * Verify that the latest {@link com.linkedin.venice.serialization.avro.AvroProtocolDefinition#KAFKA_MESSAGE_ENVELOPE}
       * version in the code base is registered in Venice backend; if not, fail fast in start phase before start writing
       * Kafka messages that Venice backend couldn't deserialize.
       */
      if (verifyLatestProtocolPresent) {
        LOGGER.info("Start verifying the latest protocols at runtime are valid in Venice backend.");
        // Discover the D2 service name for the system store
        // does not return anything useful in case of discoveryUrl
        // e.g. cannot get venice router url
        
        //D2ServiceDiscoveryResponse sysStoreDiscoveryResponse = (D2ServiceDiscoveryResponse) controllerRequestWithRetry(
        //        () -> ControllerClient.discoverCluster(discoveryUrl.get(), storeName, sslFactory, 1),
        //        2);

        String kafkaMessageEnvelopSchemaSysStore = AvroProtocolDefinition.KAFKA_MESSAGE_ENVELOPE.getSystemStoreName();
        ClientConfig clientConfigForKafkaMessageEnvelopeSchemaReader =
                ClientConfig.defaultGenericClientConfig(kafkaMessageEnvelopSchemaSysStore);
        clientConfigForKafkaMessageEnvelopeSchemaReader.setVeniceURL("http://venice-router:7777");

        SchemaReader kafkaMessageEnvelopeSchemaReader =
                ClientFactory.getSchemaReader(clientConfigForKafkaMessageEnvelopeSchemaReader, null);
        new SchemaPresenceChecker(kafkaMessageEnvelopeSchemaReader, AvroProtocolDefinition.KAFKA_MESSAGE_ENVELOPE)
                .verifySchemaVersionPresentOrExit();
        LOGGER.info("Successfully verified the latest protocols at runtime are valid in Venice backend.");
      }

      transportClient = new HttpTransportClient(discoveryUrl.get());
    } else {
```

Is there a way to discover router url using controller discovery url or I have to add/pass the router url?

Reply:
I tried to find a way to do this and gave up, decided to pass the router url as an extra parameter